### DISCUSSION TITLE
Load directed graph in fewer queries

### DISCUSSION THREAD

**Discussion by Lahaije** [sqlalchemy/sqlalchemy#10394] [Usage Questions] [answered]
I have a directed graph structure in my database model and need to load many ( > 10.000) records. The database is not on the same server as the code, causing a network delay of about 2 ms per query.

My database model is similar to https://docs.sqlalchemy.org/en/20/_modules/examples/graphs/directed_graph.html

I want to load a list of nodes and for each node, I need all edges.

This snippet is more or less the code I would like to execute.
```
node_list = [1, 2, 3, ... , 1000]
for node in session.query(Node).filter(Node.id.in_(node_list)).all():
  print(f'{node} , {node.higher_neighbors()} , {node.lower_neighbors()}')
```

For 1000 records, this results in 1 query to load the 1000 nodes, and then 2 queries for each node to load the neighbors. (2N + 1) queries in total. With a cost of 2 ms per query, this code takes about 4 seconds more than necessary when loading 1000 records.

How should I change the model and/or query to load this data in a single, or a few queries?

This is my actual code. I originally used an association proxy, but this makes no difference in the number of queries compared to the SQLAlchemy graph example. A Postgres database is used.

```
class Job(Base):
    __tablename__ = 'job'

    id: Mapped[int] = mapped_column(primary_key=True)
    job_name: Mapped[str] = mapped_column(String(256))

    parents = association_proxy("job_parents", "parent")
    children = association_proxy("job_children", "child")

    logs: Mapped[List[JobLog]] = relationship(order_by=TaskLog.id)

    def add_child(self, child, relation):
        JobRelation(parent=self, child=child, relation=relation)

    def add_parent(self, parent, relation):
        JobRelation(parent=parent, child=self, relation=relation)

class JobRelation(Base):
    __tablename__ = 'job_relation'
    parent_id: Mapped[int] = mapped_column(ForeignKey('job.id'), primary_key=True)
    child_id: Mapped[int] = mapped_column(ForeignKey('job.id'), primary_key=True)
    relation: Mapped[str] = mapped_column(String(256))

    parent = relationship(Job, foreign_keys=parent_id, backref="job_children")
    child = relationship(Job, foreign_keys=child_id, backref="job_parents")

    def __init__(self, parent: Job, child: Job, relation: str):
        self.parent = parent
        self.child = child
        self.relation = relation


job_list = json['job_list']
data = {}

for job in session.query(Job).filter(job.id.in_(job_list)).all():
    data[job.id] = {
        'children' = [child.id for child in job.children]
        'job_name' = job.job_name
        'logs' = job.logs
        'parents' = [parent.id for parent in job.parents]
```

**Comment by zzzeek** [comment, marked as answer]
typically, to reduce queries you want to use [eager loading techniques](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html), where the most general purpose for one-to-many loads is the [selectinload](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#select-in-loading).   it wont necessarily continue loading arbitrarily deep for the two class setup you have, you would need to indicate selectinload() calls for each recursive level of loading.     the query you show above only goes one level deep though so it should ideally bring the number of queries down to about three if you did this:


```py
for job in session.query(Job).filter(job.id.in_(job_list)).options(
     selectinload(Job.job_parents).joinedload(JobParent.parent), 
     selectinload(Job.job_children).joinedload(JobParent.child)
).all():
```

**Comment by Lahaije** [comment]
@zzzeek , thank you for your answer. The time to load and render a graph with 20k records dropped from 110 seconds to 28 seconds.
I did look into the loading techniques and experimented with setting the lazy='selectin' at Mapping Time, but I was unable to get it to function. I guess association_proxy does not support the lazy option at mapping Time.

Since the documentation of the relationships never mentions how to handle association_proxy I think I was unable to figure out how to enable the selectinload for my model.

**Reply by zzzeek** [reply]
loader options have to be configured independently of assocaition proxies.  pretend the association proxy attributes arent there and define the loading behavior you want in terms of the underlying relationships.

**Reply by Lahaije** [reply]
This means I can specify the loading technique in the model like
```
class JobRelation(Base):
    parent = relationship(Job, foreign_keys=parent_id, backref="job_children", lazy="joined")
    child = relationship(Job, foreign_keys=child_id, backref="job_parents", lazy="joined")
```
And rework the query like
```
for job in session.query(Job).filter(job.id.in_(job_list)).options(
     selectinload(Job.job_parents),
     selectinload(Job.job_children)
).all():
```

If I understand correctly, the query instructs loading of the 'association_proxy' object, while the model now specifies what loading technique should be used.

**Reply by zzzeek** [reply]
yes, the association proxy attribute has never been fully fleshed out as a first class ORM path that can be used in options and such...it was always meant as something to be used at the instance level.  the query-level behaviors of associaiotn proxies were bolted on later and they continue to be not really comprehensive.
